servicetalk-loadbalancer/src/main/java/io/servicetalk/loadbalancer/Host.java

```diff
@@ -15,583 +15,77 @@
-                                    // There is no risk for StackOverflowError because result of each connection
-                                    // attempt will be invoked on IoExecutor as a new task.
-                                    .retryWhen(retryWithConstantBackoffDeltaJitter(
-                                            cause -> {
-                                                LOGGER.debug("{}: health check failed for {}.",
-                                                        host.lbDescription, host, cause);
-                                                return true;
-                                            },
-                                            host.healthCheckConfig.healthCheckInterval,
-                                            host.healthCheckConfig.jitter,
-                                            host.healthCheckConfig.executor)))
-                            .flatMapCompletable(newCnx -> {
-                                if (host.addConnection(newCnx, this)) {
-                                    LOGGER.info("{}: health check passed for {}, marked this " +
-                                                    "host as ACTIVE for the selection algorithm.",
-                                            host.lbDescription, host);
-                                    return completed();
-                                } else {
-                                    // This happens only if the host is closed, no need to mark as healthy.
-                                    assert host.connState.state == State.CLOSED;
-                                    LOGGER.debug("{}: health check passed for {}, but the " +
-                                                    "host rejected a new connection {}. Closing it now.",
-                                            host.lbDescription, host, newCnx);
-                                    return newCnx.closeAsync();
-                                }
-                            })
-                            // Use onErrorComplete instead of whenOnError to avoid double logging of an error inside
-                            // subscribe(): SimpleCompletableSubscriber.
-                            .onErrorComplete(t -> {
-                                LOGGER.error("{}: health check terminated with " +
-                                        "an unexpected error for {}. Marking this host as ACTIVE as a fallback " +
-                                        "to allow connection attempts.", host.lbDescription, host, t);
-                                host.markHealthy(this);
-                                return true;
-                            })
-                            .subscribe());
-        }
+    /**
+     * Whether the host is both considered active by service discovery and healthy by the failure
+     * detection mechanisms.
+     */
+    boolean isActiveAndHealthy();
```

Comment:
It's not a comment for this PR but the overall effort.
I think we can provide a richer API that closer resembles what a control plane could also advertise about an endpoint (ie. `Condition` enum -- Terminating, NotReady, Ready) and a separate binary data plane status (healthy / unhealthy)

Reply:
+1. I had some thoughts about already but agree that we should change it in it's own PR.

Reply:
Agree as well. To add some context from xDS/EDS we definitely have this information available. "Health" from the perspective of the protocol is richer than a binary decision so it could be an interesting exercise understanding if we care.